### DISCUSSION TITLE
404 due to redirect

### DISCUSSION THREAD

**Discussion by lazarusA** [vuejs/vitepress#3722] [Q&A]
Hello again. I'm looking for some advice or hints on the following issue.
Our workflow depends on vitepress and we are setting `cleanUrls: true` if that's relevant. 

The issue: 

https://luxdl.github.io/DocumenterVitepress.jl/stable/ (show the right landing page, flickers and then gives 404 error)
https://luxdl.github.io/DocumenterVitepress.jl/dev/  (this one works) and all versioned options also work. Only this redirect is the issue. 

Maybe also relevant the corresponding redirect files:

https://github.com/LuxDL/DocumenterVitepress.jl/blob/gh-pages/index.html
https://github.com/LuxDL/DocumenterVitepress.jl/blob/gh-pages/stable

and ideas or suggestions will be very helpful.

**Comment by brc-dd** [comment]
GitHub pages don't support cleanUrls for that particular case. You'll need to use some other hosting provider if you want to avoid that. (https://github.com/vuejs/vitepress/pull/3468#issuecomment-1896430259 -- although in your case it's not changing url 👀 )

**Reply by lazarusA** [reply]
But, we have other use cases where the stable redirect points to the right version:  see: https://asinghvi17.github.io/GeometryOps.jl/stable/

**Reply by brc-dd** [reply]
Ah, right. Your case appears to be different than that because you do have `/stable/index.html` 👀  I'll take a look.

**Reply by brc-dd** [reply]
How is that site being built? Do I need to install julia or something?

**Reply by lazarusA** [reply]
unfortunately is not pure vitepress, hence Julia is necessary in order to build the docs locally.

**Reply by lazarusA** [reply]
Once, Julia is installed, simple going into the docs directory and doing in the Julia-repl should create the folders.

```Julia
Julia>] # type this to go into the pkg mode
```
```
pkg> activate .
```
then 
```
pkg> instantiate # should install all things
```
and then do

```Julia
Julia> include("make.jl")
```
and that should be it.

then in another terminal window inside the docs folder do the usual

```sh
docs> npm run docs:dev
```

**Reply by brc-dd** [reply]
It's generating stuff like this:

<img width="316" alt="image" src="https://github.com/vuejs/vitepress/assets/40380293/78e52606-f31b-4391-9cbd-4e82d934782f">

\
How to generate the directory same as your gh-pages branch? Also how to pass additional flags / environment variables to vitepress?

**Reply by lazarusA** [reply]
flags/env vars: we still can use the config file for that: 

https://github.com/LuxDL/DocumenterVitepress.jl/blob/19335ea58a79c311d1c14c35c13b14cf7814fd37/docs/src/.vitepress/config.mts

**Reply by asinghvi17** [reply]
In this case, what happens is that the Julia script detects when the CI environment variables are enabled and auto-runs the equivalent of:
```
rm -rf .documenter/
mv final_site/* ./*
rm -rf final_site
```

For convenience I've uploaded these two trees:
Site tree which does work when stable symlinks to it:
[GeometryOps.tar.gz](https://github.com/vuejs/vitepress/files/14807851/GeometryOps.tar.gz)
[SwarmMakie.tar.gz](https://github.com/vuejs/vitepress/files/14807868/SwarmMakie.tar.gz)
Site tree which does not work when stable symlinks to it:
[DocumenterVitepress.tar.gz](https://github.com/vuejs/vitepress/files/14807862/DocumenterVitepress.tar.gz)

These are all Markdown site trees, you can use this file as the package.json:
[package.json](https://github.com/vuejs/vitepress/files/14807880/package.json)

Thanks for the help!

**Reply by asinghvi17** [reply]
Also, environment variables should be picked up by the npm running in the function (we invoke it explicitly from a binary which is automatically downloaded at package install).

**Reply by brc-dd** [reply]
Something seems wrong with the redirect. Its redirecting https://asinghvi17.github.io/GeometryOps.jl/stable/ to https://asinghvi17.github.io/GeometryOps.jl/v0.0.1/ . But for your repo it's not changing the URL.

**Reply by brc-dd** [reply]
The files you shared still don't have versioned docs. I wanna know how to build that stable symlink and v0.0.14 directory. If you can do it locally, can you try building with `DEBUG=true` and sharing logs from browser console.

**Reply by asinghvi17** [reply]
If you set the environment variables `GITHUB_REPOSITORY="LuxDL/DocumenterVitepress.jl" GITHUB_EVENT_NAME="push" GITHUB_REF="v0.0.14" GITHUB_TOKEN=1` and then run `docs/make.jl` in the repo, it should give you a tagged build!

**Reply by asinghvi17** [reply]
I tried building locally with DEBUG=true in the environment, and this is what I saw in Google Chrome (site behaving correctly):
<img width="999" alt="Screenshot 2024-03-29 at 3 28 35 PM" src="https://github.com/vuejs/vitepress/assets/32143268/5a8bf1e0-4897-46bf-b343-ef61f309088e">

**Reply by brc-dd** [reply]
You're running site in dev mode. You'll need to serve build output to get logs there.

> it should give you a tagged build!

Nope, tried and it gives me the same stuff 👀

**Reply by asinghvi17** [reply]
Oops - just tried running this locally, and it seems that the doc generation package wants a few more environment variables.

`GITHUB_REPOSITORY="LuxDL/DocumenterVitepress.jl" GITHUB_EVENT_NAME="push" GITHUB_REF="refs/tags/v0.0.14" GITHUB_ACTOR="me" GITHUB_TOKEN=1` are the variables I ended up with.  Here's the built version of the docs with that:

[DocumenterVitepress-v0.0.14.tar.gz](https://github.com/vuejs/vitepress/files/14808061/DocumenterVitepress-v0.0.14.tar.gz)

What's the best way to serve build output to the browser?  I tried opening the generated HTML file directly, but that gave me a malformed result.

**Reply by brc-dd** [reply]
> DocumenterVitepress-v0.0.14.tar.gz

This too doesn't have structure similar to your gh-pages branch:

<img width="664" alt="image" src="https://github.com/vuejs/vitepress/assets/40380293/3f644a48-e718-44c7-8066-0f1011ba27fd">

Can you just set DEBUG on your CI and deploy once?

> What's the best way to serve build output to the browser?

You'll need to rename your output dir to your base (`DocumenterVitepress.jl`) and go to its parent and run `npx serve` there.

**Reply by asinghvi17** [reply]
Ah this is the source code which you would run `npm run docs:build` on.

In a few minutes a preview build with DEBUG=true should be up at https://luxdl.github.io/previews/PR88.  I will also deploy a tagged build asap!

**Reply by brc-dd** [reply]
This is the commit that broke it - https://github.com/LuxDL/DocumenterVitepress.jl/commit/27dc4e62e40018448bf324c25309059ec46f8770 - not sure why though 👀

**Reply by lazarusA** [reply]
Ok. After more debugging. We know that the failure is do to:

- an `Hydration error` (it might be do to the vue component we have on the sites not working, failing to 404). But, not a 100%, we have site without it, with the same error (https://rafaqz.github.io/DimensionalData.jl/stable/) but same redirect logic.
- and the issue happens when the re-direct is activated. 

[here](https://github.com/LuxDL/DocumenterVitepress.jl/blob/6979a763b9a7daa4e336de9c4dbdc13e44d4b424/docs/src/.vitepress/theme/index.ts#L10) This is how we include the component.

Apparently, a possible solution is to use [ClientOnly](https://github.com/vuejs/vitepress/blob/49ddb1f120f1c66c6c8d1a7e388e2f49bdf1c552/src/client/app/components/ClientOnly.ts), or maybe this https://vite-plugin-ssr.com/hydration-mismatch#example but not sure how should I modified my component calling. 

Or, is there another known solution?

Thanks again, for all the previous tips already given.

**Reply by brc-dd** [reply]
That component should work fine with SSR. It doesn't have any browser-specific code. Something else is causing the issue.
